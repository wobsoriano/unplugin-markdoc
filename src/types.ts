import type Markdoc from '@markdoc/markdoc'

export type Options = Parameters<typeof Markdoc.transform>['1']
