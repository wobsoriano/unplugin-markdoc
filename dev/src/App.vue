<script setup lang="ts">
import Markdoc from '@markdoc/markdoc'
import html from './contents/doc.md?html'

// const html = Markdoc.renderers.html(content)
</script>

<template>
  <div v-html="html" />
</template>
