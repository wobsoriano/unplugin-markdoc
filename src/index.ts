import type { Plugin } from 'vite'
import { dataToEsm } from '@rollup/pluginutils'
import Markdoc from '@markdoc/markdoc'

export type Options = Parameters<typeof Markdoc.transform>['1']

const mdExtRE = /\.(md)$/i

export function transformMarkdown(code: string, options?: Options) {
  const ast = Markdoc.parse(code)
  const content = Markdoc.transform(ast, options)

  return content
}

export default function (options?: Options): Plugin {
  return {
    name: 'vite-plugin-markdoc',
    enforce: 'pre',
    transform(code, id) {
      if (!mdExtRE.test(id))
        return null

      const esm = transformMarkdown(code, options)

      return {
        code: dataToEsm(esm, {
          preferConst: true,
          namedExports: true,
        }),
        map: { mappings: '' },
      }
    },
  }
}
