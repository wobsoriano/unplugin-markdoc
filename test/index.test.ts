import { readFile } from 'fs/promises'
import { expect, test } from 'vitest'
import { transformMarkdown } from '../src'

test('Correct markdown to tree-shakable ES Module imports', async () => {
  const content = await readFile('test/content.md', 'utf-8')
  const result = transformMarkdown(content)
  expect(result).toMatchSnapshot()
})
