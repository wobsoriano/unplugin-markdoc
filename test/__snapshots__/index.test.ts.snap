// Vitest Snapshot v1

exports[`Correct markdown to tree-shakable ES Module imports 1`] = `
Tag {
  "$$mdtype": "Tag",
  "attributes": {},
  "children": [
    Tag {
      "$$mdtype": "Tag",
      "attributes": {
        "id": "overview",
      },
      "children": [
        undefined,
        " ",
      ],
      "name": "h1",
    },
    Tag {
      "$$mdtype": "Tag",
      "attributes": {},
      "children": [
        "Markdoc is a Markdown-based syntax and toolchain for creating custom documentation sites. Stripe created Markdoc to power ",
        Tag {
          "$$mdtype": "Tag",
          "attributes": {
            "href": "http://stripe.com/docs",
          },
          "children": [
            "our public docs",
          ],
          "name": "a",
        },
        ".",
      ],
      "name": "p",
    },
    Tag {
      "$$mdtype": "Tag",
      "attributes": {},
      "children": [
        "Markdoc is open-source—check out its ",
        Tag {
          "$$mdtype": "Tag",
          "attributes": {
            "href": "http://github.com/markdoc/markdoc",
          },
          "children": [
            "source",
          ],
          "name": "a",
        },
        " to see how it works.",
      ],
      "name": "p",
    },
    Tag {
      "$$mdtype": "Tag",
      "attributes": {},
      "children": [
        "How is Markdoc different?",
      ],
      "name": "h2",
    },
    Tag {
      "$$mdtype": "Tag",
      "attributes": {},
      "children": [
        "Markdoc uses a fully declarative approach to composition and flow control, where other solutions… ",
        Tag {
          "$$mdtype": "Tag",
          "attributes": {
            "href": "/docs/overview",
          },
          "children": [
            "Read more",
          ],
          "name": "a",
        },
        ".",
      ],
      "name": "p",
    },
    Tag {
      "$$mdtype": "Tag",
      "attributes": {},
      "children": [
        "Next steps",
      ],
      "name": "h2",
    },
    Tag {
      "$$mdtype": "Tag",
      "attributes": {},
      "children": [
        Tag {
          "$$mdtype": "Tag",
          "attributes": {},
          "children": [
            Tag {
              "$$mdtype": "Tag",
              "attributes": {
                "href": "/docs/getting-started",
              },
              "children": [
                "Install Markdoc",
              ],
              "name": "a",
            },
          ],
          "name": "li",
        },
        Tag {
          "$$mdtype": "Tag",
          "attributes": {},
          "children": [
            Tag {
              "$$mdtype": "Tag",
              "attributes": {
                "href": "/docs/syntax",
              },
              "children": [
                "Explore the syntax",
              ],
              "name": "a",
            },
          ],
          "name": "li",
        },
      ],
      "name": "ul",
    },
  ],
  "name": "article",
}
`;
